web/packages/teleterm/src/ui/hooks/useAppState.ts

```diff
@@ -0,0 +1,108 @@
+/**
+ * Teleport
+ * Copyright (C) 2024 Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { useCallback, useState } from 'react';
+
+import { useAppContext } from 'teleterm/ui/appContextProvider';
+import { StatePersistenceState } from 'teleterm/ui/services/statePersistence';
+
+/**
+ * useAppState is like useState, but it persists the state to app_state.json under the given key.
+ *
+ * IMPORTANT: Currently, useAppState doesn't propagate changes across several callsites. That is, if
+ * two callsites use the same key, calling setState in one component will not cause the other
+ * component to update.
+ *
+ * This will _not_ work as expected:
+ *
+ * const Counter = () => {
+ *   const [count, setCount] = useAppState('count', 0);
+ *
+ *   return (
+ *     <div>
+ *       {count}
+ *       <button onClick={() => setCount(count + 1)}>Increase</button>
+ *     </div>
+ *   );
+ * }
+ *
+ * () => {
+ *   return (
+ *     <>
+ *       <Counter />
+ *       <Counter />
+ *     </>
+ *   );
+ * }
+ *
+ * However, this will work as expected:
+ *
+ * @example
+ * const Counter = ({count, setCount}) => {
+ *   return (
+ *     <div>
+ *       {count}
+ *       <button onClick={() => setCount(count + 1)}>Increase</button>
+ *     </div>
+ *   );
+ * }
+ *
+ * () => {
+ *   const [count, setCount] = useAppState('count', 0);
+ *
+ *   return (
+ *     <>
+ *       <Counter count={count} setCount={setCount} />
+ *       <Counter count={count} setCount={setCount} />
+ *     </>
+ *   );
+ * }
+ */
+export function useAppState<
```

Comment:
I have a feeling that this name could be more specific, for example, `usePersistedState` or `useAppPersistedState`. 
Without checking the docs it is quite hard to guess what kind of state it updates (maybe it's a runtime state that lives in `AppContext` services?).

Reply:
Renamed to `usePersistedState` after a short discussion in DMs.

Reply:
…and renamed `useImmutableStore` to `useStoreSelector`, because as Grzegorz pointed out to me, the "immutable" part is just an implementation details.